### PR TITLE
Add YOLO class.

### PR DESCRIPTION
**Pull request by kartikdutt18**
Hey everyone,
This PR aims to YOLO class to the models repo. The following is the TO-DO list:

- [x] Add PreProcessor Function (Convert DataLoader annotations to YOLO targets). (I'll add this by 10.07.20)

- [x] Add Detection stage for the YOLO class (Where loss is calculated, feature maps are converted to understandable format).

- [x] Add YOLO model architecture.

### CODE REVIEW

**Comment by kartikdutt18** [comment]
The build failure occurs because it doesn't have the newer batchnorm yet.

**Comment by zoq** [comment]
I guess with https://github.com/mlpack/mlpack/pull/2474 merged, this should build now? Let's rerun the tests.

**Comment by kartikdutt18** [comment]
> I guess with mlpack/mlpack#2474 merged, this should build now? Let's rerun the tests.

Yes, it builds now.

**Comment by saksham189** [comment]
Can you fix the style check build?

**Comment by kartikdutt18** [comment]
Sure, I will resolve it with the next commit. Thanks for pointing it out.

File: models/yolo/yolo_impl.hpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    weights(weights)
+{
+  std::set<string> supportedVersion({"v1-tiny"});
+  mlpack::Log::Assert(supportedVersion.count(YOLOVersion),
+      "Unsupported YOLO version. Trying to find :", YOLOVersion);
+
+  if (weights == "voc")
+  {
+    // Download weights here.
+    LoadModel("./../weights/YOLO/yolo" + YOLOVersion + "_voc.bin");
```
[the remark this answers is not in the document]
Comment (kartikdutt18) [reply models/yolo/yolo_impl.hpp]: I have added a simple tests in this PR. I haven't added tests for weights because of mlpack/mlpack#2561. I think we can merge this PR since the model is working and any changes would be internal implementation in mlpack. To see if the model is working correctly you can run the following commands :
```
git clone https://github.com/kartikdutt18/mlpack-PyTorch-Weight-Translator.git
cd mlpack-PyTorch-Weight-Translator
./run.sh yolov1_tiny
```
With this you will see the model weights being created and transferred into the YOLO model and being used for prediction. You can see that the iou of predictions made by PyTorch model and YOLO model are same i.e. 1.0. Kindly let me know if you face any issues. If this makes sense, I think we can merge this PR and then later add a PR with tests for weights once mlpack/mlpack#2561 is resolved.
(Similar comment from DarkNet).

[the remark this answers is not in the document]
Comment (kartikdutt18) [reply models/yolo/yolo_impl.hpp]: With this we should also merge #29 and #25 as well.

**Review by KimSangYeon-DGU — APPROVED** [review]
No further comments from my side.

File: models/yolo/yolo_impl.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (KimSangYeon-DGU) [reply models/yolo/yolo_impl.hpp]: @kartikdutt18 Rather than creating a weight file, how about save it to the `weight` directory for making it ready-for-use?

File: models/yolo/yolo.hpp [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+  /**
+   * YOLO constructor intializes input shape and number of classes.
+   *
+   * @param inputChannels Number of input channels of the input image.
+   * @param inputWidth Width of the input image.
+   * @param inputHeight Height of the input image.
+   * @param yoloVersion Version of YOLO model.
+   * @param numClasses Optional number of classes to classify images into,
+   *                   only to be specified if includeTop is  true.
```
Comment (KimSangYeon-DGU) [inline models/yolo/yolo.hpp:64]:
```suggestion
   * @param numClasses Optional number of classes to classify images into,
   *     only to be specified if includeTop is  true.
```

File: models/yolo/yolo.hpp [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+       const std::string& weights = "none",
+       const bool includeTop = true);
+
+  /**
+   * YOLO constructor intializes input shape and number of classes.
+   *
+   * @param inputShape A three-valued tuple indicating input shape.
+   *                   First value is number of Channels (Channels-First).
+   *                   Second value is input height.
+   *                   Third value is input width.
```
Comment (KimSangYeon-DGU) [inline models/yolo/yolo.hpp:88]: Please use 4 spaces in a new line.

```suggestion
   * @param inputShape A three-valued tuple indicating input shape.
   *     First value is number of Channels (Channels-First).
   *              ...
```

File: models/yolo/yolo.hpp [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+   * @param yoloVersion Version of YOLO model.
+   * @param numClasses Optional number of classes to classify images into,
+   *                   only to be specified if includeTop is  true.
```
Comment (KimSangYeon-DGU) [inline models/yolo/yolo.hpp:91]:
```suggestion
   * @param numClasses Optional number of classes to classify images into,
   *     only to be specified if includeTop is  true.
```

File: models/yolo/yolo.hpp [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+   * @param numBoxes Number of bounding boxes per image.
+   * @param featureShape A twp-valued tuple indicating width and height of output feature
+   *                     map.
```
Comment (KimSangYeon-DGU) [inline models/yolo/yolo.hpp:94]:
```suggestion
   * @param featureShape A twp-valued tuple indicating width and height of output feature
   *     map.
```

File: models/yolo/yolo.hpp [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+  //! Save weights for the model.
+  void SaveModel(const std::string& filePath);
+
+ private:
+  /**
+   * Adds Convolution Block.
+   *
+   * @tparam SequentialType Layer type in which convolution block will
+   *                        be added.
```
Comment (KimSangYeon-DGU) [inline models/yolo/yolo.hpp:119]:
```suggestion
   * @tparam SequentialType Layer type in which convolution block will
   *     be added.
```

File: models/yolo/yolo.hpp [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+   * @param kernelWidth Width of the filter/kernel.
+   * @param kernelHeight Height of the filter/kernel.
+   * @param strideWidth Stride of filter application in the x direction.
+   * @param strideHeight Stride of filter application in the y direction.
+   * @param padW Padding width of the input.
+   * @param padH Padding height of the input.
+   * @param batchNorm Boolean to determine whether a batch normalization
+   *                  layer is added.
+   * @param baseLayer Layer in which Convolution block will be added, if
+   *                  NULL added to YOLO FFN.
```
Comment (KimSangYeon-DGU) [inline models/yolo/yolo.hpp:132]:
```suggestion
   * @param batchNorm Boolean to determine whether a batch normalization
   *     layer is added.
   * @param baseLayer Layer in which Convolution block will be added, if
   *     NULL added to YOLO FFN.
```

File: models/yolo/yolo.hpp [hunk 8]
```diff
@@ -0,0 +1,255 @@
...
+    else
+      yolo.Add(bottleNeck);
+  }
+
+  /**
+   * Adds Pooling Block.
+   *
+   * @param factor The factor by which input dimensions will be divided.
+   * @param type One of "max" or "mean". Determines whether add mean pooling
+   *             layer or max pooling layer.
```
Comment (KimSangYeon-DGU) [inline models/yolo/yolo.hpp:176]:
```suggestion
   * @param type One of "max" or "mean". Determines whether add mean pooling
   *     layer or max pooling layer.
```

**Review by KimSangYeon-DGU — COMMENTED** [review]
Left some comments about styles

File: models/yolo/yolo_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+>::LoadModel(const std::string& filePath)
+{
+  data::Load(filePath, "yolo" + yoloVersion, yolo);
+  Log::Info << "Loaded model." << std::endl;
+}
+
+template<
+     typename OutputLayerType,
+     typename InitializationRuleType
+>
```
Comment (KimSangYeon-DGU) [inline models/yolo/yolo_impl.hpp:154]: Please use 4 spaces.

```suggestion
template<
    typename OutputLayerType,
    typename InitializationRuleType
>
```

File: models/yolo/yolo_impl.hpp [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+      weights,
+      includeTop)
+{
+  // Nothing to do here.
+}
+
+template<
+     typename OutputLayerType,
+     typename InitializationRuleType
+>
```
Comment (KimSangYeon-DGU) [inline models/yolo/yolo_impl.hpp:71]:
```suggestion
template<
    typename OutputLayerType,
    typename InitializationRuleType
>
```

**Review by KimSangYeon-DGU — CHANGES_REQUESTED** [review]

File: models/yolo/yolo_impl.hpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kartikdutt18) [reply models/yolo/yolo_impl.hpp]: This would face the same problem as DarkNet since this also uses BatchNorm Layer which has a serialization issue.

File: models/yolo/yolo_impl.hpp (hunk 10 above)
Comment (kartikdutt18) [reply models/yolo/yolo_impl.hpp]: Sorry about the same style errors repeated in different PRs. I think I fixed them all. Thanks for pointing them out.

**Review by KimSangYeon-DGU — APPROVED** [review]
No further comments from my side.

**Comment by KimSangYeon-DGU** [comment]
Thanks for another great contribution :+1:

**Comment by kartikdutt18** [comment]
Awesome, Thanks a lot @KimSangYeon-DGU and @saksham189 for all the reviews, suggestions and the help.
